modules/core/src/main/java/org/apache/ignite/internal/processors/cache/distributed/dht/preloader/GridDhtPartitionSupplier.java

```diff
@@ -475,6 +478,11 @@ else if (iter.isPartitionMissing(p)) {
                 log.info("Finished supplying rebalancing [" + supplyRoutineInfo(topicId, nodeId, demandMsg) + "]");
         }
         catch (Throwable t) {
+            boolean fallbackToFullRebalance = (iter == null) ?
```

Comment:
This code looks fragile. Do we not want to do fallback if this is surely a WAL error, like IO error ?

Reply:
Do you mean that we need a special exception (something like as follows: `WalHistoryIteratorInitializationException`) and use it for this case?

Reply:
Introduced `IgniteHistoricalIteratorException`